coffee/rewriter.py

```diff
@@ -407,124 +420,109 @@ def licm(self):
             self.extracted = False
             self.counter += 1
 
-            for all_deps, expr in expr_dep.items():
+            for all_deps, exprs in expr_dep.items():
                 # -1) Filter dependencies that do not pertain to the expression
-                dep = tuple(d for d in all_deps if d in real_deps)
-
-                # 0) The invariant statements go in the closest outer loop to
-                # dep[-1] which they depend on, and are wrapped by a loop wl
-                # iterating along the same iteration space as dep[-1].
-                # If there's no such an outer loop, they fall in the header,
-                # provided they are within a perfect loop nest (otherwise,
-                # dependencies may be broken)
-                outermost_loop = expr_loops[0]
-                is_outermost_perfect = is_perfect_loop(outermost_loop)
+                dep = tuple(d for d in all_deps if d in self.expr_deps)
+
+                # 0) Determine where it is safe to hoist the statements
                 if len(dep) == 0:
-                    place, wl = self.header, None
-                    next_loop = outermost_loop
-                elif len(dep) == 1 and is_outermost_perfect:
-                    place, wl = self.header, dict_expr_loops[dep[0]]
-                    next_loop = outermost_loop
-                elif len(dep) == 1 and not is_outermost_perfect:
-                    place, wl = dict_expr_loops[dep[0]].children[0], None
-                    next_loop = od_find_next(dict_expr_loops, dep[0])
+                    place, wrap_loop = self.header, None
+                    next_loop = expr_outermost_loop
+                elif len(dep) == 1 and is_expr_outermost_perfect:
+                    place, wrap_loop = self.header, expr_dims_loops[dep[0]]
+                    next_loop = expr_outermost_loop
+                elif len(dep) == 1 and len(expr_dims_loops) > 1:
+                    place, wrap_loop = expr_dims_loops[dep[0]].children[0], None
+                    next_loop = od_find_next(expr_dims_loops, dep[0])
+                elif len(dep) == 1:
+                    place, wrap_loop = expr_dims_loops[dep[0]].children[0], None
+                    next_loop = place.children[-1]
                 else:
-                    dep_block = dict_expr_loops[dep[-2]].children[0]
-                    place, wl = dep_block, dict_expr_loops[dep[-1]]
-                    next_loop = od_find_next(dict_expr_loops, dep[-2])
+                    dep_block = expr_dims_loops[dep[-2]].children[0]
+                    place, wrap_loop = dep_block, expr_dims_loops[dep[-1]]
+                    next_loop = od_find_next(expr_dims_loops, dep[-2])
 
                 # 1) Remove identical sub-expressions
-                expr = dict([(str(e), e) for e in expr]).values()
+                exprs = dict([(str(e), e) for e in exprs]).values()
```

Comment:
I need filtering to happen on the string representation to delete duplicates, then I pick the expressions